warehouse/attestations/services.py

```diff
@@ -0,0 +1,249 @@
+            raise AttestationUploadError(
+                f"Unsupported publisher: {publisher.publisher_name}"
+            )
+
+
+def _build_provenance(
+    request: Request, file: File, attestations: list[Attestation]
+) -> DatabaseProvenance:
+    try:
+        publisher: Publisher = _publisher_from_oidc_publisher(request.oidc_publisher)
+    except AttestationUploadError as exc:
+        sentry_sdk.capture_message(
+            f"Unsupported OIDCPublisher found {request.oidc_publisher.publisher_name}"
+        )
+        raise exc
+
+    attestation_bundle = AttestationBundle(
+        publisher=publisher,
+        attestations=attestations,
+    )
+
+    provenance = Provenance(attestation_bundles=[attestation_bundle]).model_dump(
+        mode="json"
+    )
+
+    db_provenance = DatabaseProvenance(
+        file=file,
+        provenance=provenance,
+        provenance_digest=hashlib.sha256(rfc8785.dumps(provenance)).hexdigest(),
+    )
+
+    return db_provenance
+
+
+def _extract_attestations_from_request(request: Request) -> list[Attestation]:
+    """
+    Extract well-formed attestation objects from the given request's payload.
+    """
+
+    metrics = request.find_service(IMetricsService, context=None)
+
+    try:
+        attestations = TypeAdapter(list[Attestation]).validate_json(
+            request.POST["attestations"]
+        )
+    except ValidationError as e:
+        # Log invalid (malformed) attestation upload
+        metrics.increment("warehouse.upload.attestations.malformed")
+        raise AttestationUploadError(
+            f"Malformed attestations: {e}",
+        )
+
+    # Empty attestation sets are not permitted; users should omit `attestations`
+    # entirely to upload without attestations.
+    if not attestations:
+        raise AttestationUploadError(
+            "Malformed attestations: an empty attestation set is not permitted"
+        )
+
+    # This is a temporary constraint; multiple attestations per file will
+    # be supported in the future.
+    if len(attestations) > 1:
+        metrics.increment("warehouse.upload.attestations.failed_multiple_attestations")
+
+        raise AttestationUploadError(
+            "Only a single attestation per file is supported",
+        )
+
+    return attestations
+
+
+@implementer(IIntegrityService)
+class NullIntegrityService:
+    def __init__(self, session):
+        warnings.warn(
+            "NullIntegrityService is intended only for use in development, "
+            "you should not use it in production due to the lack of actual "
+            "attestation verification.",
+            InsecureIntegrityServiceWarning,
+        )
+        self.db = session
+
+    @classmethod
+    def create_service(cls, _context, request):
+        return cls(session=request.db)
+
+    def parse_attestations(
+        self, request: Request, _distribution: Distribution
+    ) -> list[Attestation]:
+        publisher: OIDCPublisher | None = request.oidc_publisher
```

Comment:
FLAG: This test is performed in the IntegrityService but was not done here and adding it highlighted some issues.